expression/builtin_compare.go

```diff
@@ -1699,6 +1727,31 @@ func (c *compareFunctionClass) refineArgs(ctx sessionctx.Context, args []Express
 	return c.refineArgsByUnsignedFlag(ctx, []Expression{finalArg0, finalArg1})
 }
 
+// see https://github.com/pingcap/tidb/issues/38361 for more details
+func (c *compareFunctionClass) refineNumericConstantCmpDatetime(ctx sessionctx.Context, args []Expression, constArg *Constant, constArgIdx int) []Expression {
+	dt, err := constArg.Eval(chunk.Row{})
+	if err != nil {
+		return args
+	}
+	sc := ctx.GetSessionVars().StmtCtx
+	var timestampDatum types.Datum
+	targetFieldType := types.NewFieldType(mysql.TypeTimestamp)
```

Comment:
Why the target type is timestamp instead of datetime?

Reply:
According to MySQL document, "If one of the arguments is a [TIMESTAMP](https://dev.mysql.com/doc/refman/8.0/en/datetime.html) or [DATETIME](https://dev.mysql.com/doc/refman/8.0/en/datetime.html) column and the other argument is a constant, the constant is converted to a timestamp before the comparison is performed"

Reply:
Check that the MySQL seems convert to datetime instead, because it can handle numeric constant that is beyond timestamp's valid value range.